[Amadeus] iP

Amadeus-Winarto opened the pull request:
# BobCat
> “This is very fun!” – Me 

## What is BobCat
BobCat is: 
- text-based
- easy to learn
-  ~~FAST~~ SUPER FAST (hopefully)

## How do I use this thing? 
To use it, follow these steps: 
1. Download it from [here](https://www.google.com)
2. Double-click it the jar file
3. Add your tasks.
4. And be **happy!** 😉

## Features
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)
- [ ] If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Amadeus-Winarto committed:
Working baseline for Level 1

Amadeus-Winarto committed:
Modified initial string

Amadeus-Winarto committed:
Changed chatbot name

Amadeus-Winarto committed:
Add memory capability

Amadeus-Winarto committed:
Support storing completion status

Amadeus-Winarto committed:
Support marking tasks as done

Amadeus-Winarto committed:
Refactored to support different tasks

Amadeus-Winarto committed:
Close scanner object

Amadeus-Winarto committed:
Added tests

Amadeus-Winarto committed:
Added exceptions

Amadeus-Winarto committed:
Added tests for exceptions

Amadeus-Winarto committed:
Added support for delete

Amadeus-Winarto committed:
Added more exceptions

Amadeus-Winarto committed:
Restructured code

Amadeus-Winarto committed:
Update tests

Amadeus-Winarto committed:
Refactor parser

Amadeus-Winarto committed:
Refactor parser further

Amadeus-Winarto committed:
Refactored exceptions to show hierarchy

Amadeus-Winarto committed:
Fixed tests

Amadeus-Winarto committed:
Minor refactoring to support saving to disk

Amadeus-Winarto committed:
Update tests

Amadeus-Winarto committed:
Merge branch 'master' into branch-Level-7

Amadeus-Winarto committed:
Fixed tests

Amadeus-Winarto committed:
Support save to disk

Amadeus-Winarto committed:
Minimum Date/Time support

Amadeus-Winarto committed:
Merge branch 'branch-Level-7'

Amadeus-Winarto committed:
Merge branch 'branch-Level-8'

Amadeus-Winarto committed:
Ignore memory files

Amadeus-Winarto committed:
Minor refactor to respect milestone requirements

Amadeus-Winarto committed:
Minor refactor to respect milestone requirements

Amadeus-Winarto committed:
Minor refactor to respect milestone requirements

Amadeus-Winarto committed:
Fixed storage parsers

Amadeus-Winarto committed:
Update junit tests

Amadeus-Winarto committed:
Minor view updates

Amadeus-Winarto committed:
Added JavaDoc

Amadeus-Winarto committed:
Fixed Naming and Layout

Amadeus-Winarto committed:
Complete style check

Amadeus-Winarto committed:
Added support for find command

Amadeus-Winarto committed:
Merge branch 'branch-A-JavaDoc'

Amadeus-Winarto committed:
Merge branch 'branch-A-CodingStandard'

Amadeus-Winarto committed:
Merge branch 'branch-Level-9'

Amadeus-Winarto committed:
Move BobCat for futher compliance to standards

Amadeus-Winarto committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

Amadeus-Winarto committed:
Setup gradle

Amadeus-Winarto committed:
Modify build.gradle

Amadeus-Winarto committed:
Add checkstyle

clyveycui commented on src/main/java/BobCat/executor/parser/CommandParser.java:
```diff
@@ -0,0 +1,9 @@
+package BobCat.executor.parser;
+
+/**
+ * Interface that all command parsers in BobCat.BobCat must obey. A <code>CommandParser</code> checks if the given command
+ * and commandArgs are appropriate for each other. Specifically, it checks for the number of given arguments.
+ */
+public interface CommandParser {
+    String[] parse(String command, String[] commandArgs);
+}
```
Interesting design decision here. May I know why did you decide to create so many different CommandParsers?

Amadeus-Winarto replied:
Hi. The decision was made due to the intention to categorise a *primitive* Command into 3 kinds: a BasicCommand which does read-only operations, a TaskCreation command which creates a new Task, as well as a TaskMutationCommand, which mutates a given Task.

The reason behind this hierarchy is to force an implementer to create commands with a clear, small job i.e. either read-only, create or mutate.  Other kinds of commands can then be created as a composition of these primitive commands, thus forcing reuse of the already written primitive commands. 

The myriad of parsers are written to implement this idea. This leads to BasicCommandParser, CreationCommandParser, and MutationCommandParser. An alternative approach (that I can think of) is to have a `Command` class, with methods `execute` and `parse`. Not too sure which one is better, if there's even a difference, so currently sticking to this approach. 

QueryParser is a general parser that calls the appropriate CommandParser. It does not implement CommandParser. The same holds for StorageParser, which parses the storage file. This is done to allow flexibility in how storage files are written.

clyveycui commented on src/main/java/BobCat/model/TaskList.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.ArrayList;
+
+public class TaskList {
+    private final ArrayList<Task> storage;
```
The variable name here should be in uppercase I believe.

Amadeus-Winarto replied:
Can I clarify that the use of uppercase here is due to the "final" keyword? Was under the impression that constants in Java refer to static final fields, not just final

clyveycui reviewed (commented):
Awesome job! Just one quick fix and should be good to go.

Amadeus-Winarto committed:
Make CheckStyle compliant

Amadeus-Winarto committed:
Fix: text-ui tests

Amadeus-Winarto committed:
Major refactoring to support GUI

Amadeus-Winarto committed:
Ignore javafx

Amadeus-Winarto committed:
Ignore javafx

Amadeus-Winarto committed:
Merge branch 'master' into branch-A-Gradle

Amadeus-Winarto committed:
Merge branch 'branch-A-Gradle'

Amadeus-Winarto committed:
Gradle final setup

Amadeus-Winarto committed:
Added gradle support

Amadeus-Winarto committed:
Merge branch 'branch-A-CheckStyle'

Amadeus-Winarto committed:
Gradle setup for JavaFX

Amadeus-Winarto committed:
Merge branch 'branch-Level-10'

SpdPnd98 commented on src/main/java/bobcat/executor/parser/QueryParser.java:
```diff
@@ -0,0 +1,46 @@
+ * sets of commands.
+ */
+public class QueryParser {
+    private static final Set<String> taskCreation = Set.of("todo", "event", "deadline");
+    private static final Set<String> taskMarking = Set.of("done", "delete");
+    private static final Set<String> basicCommand = Set.of("list", "find", "bye");
```
Should these be all caps and snake case?

SpdPnd98 reviewed (commented):
LGTM! Code looks well organised!

CrownKira commented on src/main/java/bobcat/exception/BobCatException.java:
```diff
@@ -0,0 +1,7 @@
+package bobcat.exception;
+
+public class BobCatException extends RuntimeException {
```
Perhaps you would like to include a header comment for this class since it is public?

CrownKira replied:
Noticed this in several other places too.

CrownKira commented on src/main/java/bobcat/executor/ExecutionUnit.java:
```diff
@@ -0,0 +1,92 @@
+
+    /**
+     * Executes the given query. ExecutionUnit will pass the query to the appropriate parser and execute the
+     * arguments according to the command recognised by the parser.
```
Should you have a separation line between subject and param?

CrownKira commented on src/main/java/bobcat/executor/ExecutionUnit.java:
```diff
@@ -0,0 +1,92 @@
+     * @return Array of strings representing information received after query is executed
+     * @throws IOException may be thrown if given storagePath to ExecutionUnit does not exist
+     */
+    public String[] executeCommand(String query) throws IOException {
```
Perhaps you can refactor this method into multiple parts? eg. extract the logic for each action (list, bye, etc)? I would recommend to keep a method to <30 LoC

CrownKira commented on src/main/java/bobcat/executor/parser/QueryParser.java:
```diff
@@ -0,0 +1,46 @@
+    private final MutationCommandParser taskMarkingParser = new MutationCommandParser();
+
+    /**
+     * Process user query and call the appropriate <code>CommandParser</code>.
```
Processes*

CrownKira replied:
calls

CrownKira commented on src/main/java/bobcat/model/TaskList.java:
```diff
@@ -0,0 +1,75 @@
+    private int index;
+
+    /**
+     * Constructor for a task list
```
Instead of starting with a noun, perhaps you could phrase it this way: Constructs an empty task list

CrownKira commented on src/main/java/bobcat/view/DialogBox.java:
```diff
@@ -0,0 +1,75 @@
+        this.getChildren().addAll(l, iv);
+    }
+
+    //    private void flip() {
```
Perhaps you would like to 'garbage collect' the code if deprecated ?

CrownKira reviewed (commented):
LGTM!

Amadeus-Winarto committed:
Refactor executeCommand method of ExecutionUnit

executeCommand was long and not easy to read.

Shifting mainLogic to a new package increases readibility while
reflecting hierarchy of classes.

Amadeus-Winarto committed:
Add assertions to methods taking in Java Object

There are several methods which takes in a Java Object that will fail if the Object is null.

Adding assertions to these methods improves code readability and
debugging process in the future.

Amadeus-Winarto committed:
Merge pull request #1 from Amadeus-Winarto/branch-A-Assertions

Add assertions to methods taking in Java Object

Amadeus-Winarto committed:
Merge branch 'master' of github.com:Amadeus-Winarto/ip

Amadeus-Winarto committed:
Merge branch 'master' into branch-A-CodeQuality

Amadeus-Winarto committed:
Merge pull request #2 from Amadeus-Winarto/branch-A-CodeQuality

Refactor executeCommand method of ExecutionUnit

Amadeus-Winarto committed:
CI attempt

Amadeus-Winarto committed:
setup actions

Amadeus-Winarto committed:
Merge branch 'branch-A-CI'

Amadeus-Winarto committed:
Make findByName() method do fuzzy search

findByName() requires input to be subset of description in TaskList.

Making findByName() do fuzzy search improves typo tolerance for users.

As a step towards implementing fuzzy search, let's make a baseline
based on Levenshtein Distance. Doing so will make subsequent implementation
easier.

Amadeus-Winarto committed:
Merge branch 'branch-C-BetterSearch'

Amadeus-Winarto committed:
Minor improvements

Amadeus-Winarto committed:
Add application name to window

Amadeus-Winarto committed:
Add docs

Amadeus-Winarto committed:
Add acknowledgement section

Amadeus-Winarto committed:
Update doc

Amadeus-Winarto committed:
Set theme jekyll-theme-hacker

Amadeus-Winarto committed:
Resolve #3

Amadeus-Winarto committed:
Merge branch 'master' of github.com:Amadeus-Winarto/ip
